DS-865 Token Implementation notes (#109)
nypublicradio/nypr-design-system: stories/10-For Developers/Token Implementation Notes.stories.mdx

import { Meta } from '@storybook/addon-docs/blocks';

<Meta title="For Developers/Token Implementation Notes" />

Tokens are placeholders for specific values in our CSS. For example, instead of specifiying text color as a dark gray, we can simply say it's the "text color" and let each theme fill in the details.

Tokens are what power Radial's themability. By building our components with tokens instead of specific values, we allow them to be easily rethemed. Tokens won't cover everything we want in a new design but they let us tune the defaults in a way that resembles configuration more than coding.

## Glossary

There are a few terms around tokens that can be confusing so let's define them up front.

**Variables:** A name and a value. Here we're usually talking about CSS custom properties, also known as CSS variables. This is the technical feature we use to implement tokens.

> **Tokens:** CSS Variables that a developer uses when implementing the base styles for a component. Not all variables are tokens. For example, a theme might have a local palette of colors that's used to set the token values. When I write some code for a theme that defines "button color" as "brand red", "button color" is a token, but "brand red" isn't.

Reviewer: ```suggestion
**Tokens:** CSS Variables that a developer uses when implementing the base styles for a component. Not all variables are tokens. For example, a theme might have a local palette of colors that's used to set the token values. When I write some code for a theme that defines "button color" as "brand red," "button color" is a token, but "brand red" isn't.
```

> **Dials:** Variables that can be set by a theme author to change the value of a single token. For example, "button color".

Reviewer: ```suggestion
**Dials:** Variables that can be set by a theme author to change the value of a single token. For example, "button color."
```

**Levers:** Variables that can be set by a theme author to change the value of multiple tokens. For example, the font scaling factor which changes all of the derived font size tokens.

## CSS Custom Properties

We chose to go with CSS custom properties over SASS variables because of several features that weren't easily replicated with SASS.

Advantages
- Preserving the variable names in source allows us to use the CSS as a source of truth in our storybook docs.
- Being able to see the variable names in the browser devtools makes debugging more transparent.

> - CSS's scoping rules makes implementing color groups using CSS variables pretty straightforward.

Reviewer: ```suggestion
- CSS's scoping rules make implementing color groups using CSS variables pretty straightforward.
```

Disadvantages:
- They have limited support in IE.

## Token Linting

We use stylelint to enforce an allowlist so that things like theme specific colors don't get added to component styles.

Right now we only apply this check to the base styles and are less strict with what goes into the site specific themes themselves. There's no benifit to saying "brand theme" can't use "brand color" directly.

## Component Specific Tokens

> TODO: Add recommendations for adding dials to components. This will probably be detailed more in it's own doc or a doc about components.

Reviewer: ```suggestion
TODO: Add recommendations for adding dials to components. This will probably be detailed more in its own doc or a doc about components.
```

    }
}

.my-header.mod-smaller {
    font-size:      var(--font-size-6);
    line-height:    var(--line-height-6);

    @include media(">=medium") {
       font-size:      var(--font-size-8);
       line-height:    var(--line-height-8);
    }
}
```

```
// DON'T DO THIS
.my-button {
  @include typeface (body, 6);
  line-height: 100%; // customized line height for buttons.
}

.my-button.mod-small {
  @include typeface (body, 4); // overwrites the customized line height. oops!
}
```

## Spacing

Radial includes three predefined spacing scales, Compact, Default, and Loose. To choose the spacing scale for a theme, the theme author can change the variables `--spacing-base` and `spacing-increment` to use a pair of these predefined variables.

Setting a theme to use compact spacing:
```
  --spacing-base:      var(--spacing-base-compact);
  --spacing-increment: var(--spacing-increment-compact);
```

Setting these levers configures the values for a range of spacing tokens, numbered `--space-1` to `--space-11`.

When creating components, developers and designers should use these tokens consistely according to the spacing guidelines.

## Colors

### Color Format

> One nice feature of SASS is its built in color functions. You can write `rgba(#ff000, 0.8)` and it will become `rgba(255, 0, 0, 0.8)`. Or more usefully you might write something like `rgba($theme-red, $hover-opacity)`.

Reviewer: ```suggestion
One nice feature of SASS is its built-in color functions. You can write `rgba(#ff000, 0.8)` and it will become `rgba(255, 0, 0, 0.8)`. Or more usefully you might write something like `rgba($theme-red, $hover-opacity)`.
```

- Supported CSS color formats won't let use rgb colors:
- Using the 8 digit hex format (#RRGGBBAA) almost works, but we can't concatenate css vars without a space in-between. `--my-green: '#229922'; --my-opacity: 'cc'; color: var(--my-green)var(--my-opacity);` doesn't work. nor does `--my-green: '#229922'; color: var(--my-green)cc;`.
-  `--my-green: '#229922'; rgba(#229922, 0.8)` also won't work in plain CSS.

We ended up storing colors as a comma separated set of rgb values: e.g. `--my-green: "34,153,34";`
This lets us do `color: rgba(var(--my-green), 0.8)` or even `color: rgba(var(--my-green), (--opacity-token))`

Since the design team uses hex colors in their designs, we include a sass function that will convert hex colors to these rgb values, to make theme files easier to parse.

`--my-green: hex2rgb("229922");` becomes `--my-green: "34,153,34"` in the compiled css.

I didn't find this until after we sorted this all out, but the first part of this blog post covers this issue and came to basically this exact same solution: https://blog.jim-nielsen.com/2019/generating-shades-of-color-using-css-variables/

### The base palette

> Because color tokens will be need to be customized for each theme, and also documented in several places we don't want to go overboard with the list of color tokens. The base set should be colors that are use widely in multiple places across our sites and provide a useful default pallete that we can use to build our components.

Reviewer: ```suggestion
Because color tokens will need to be customized for each theme and documented in several places, we don't want to go overboard with the list of color tokens. The base set should include colors that are used widely in multiple places across our sites and provide a useful default palette that we can use to build our components.
```

#### Primary Theme Colors

The guidelines for where to use these within components are still a little unclear, but basically these are the brand colors for the site.

#### Background Colors

The main background color and some alternates for special classes.

You can create mod classes or overrides to use these alternative background colors.

```
//theme.overrides.components.popup.scss
.popup.mod-muted {
  // this theme provides an alternative style for popups with a less bright background color
  --color-background: var(--color-background-muted);
}

//theme.overrides.components.footer.scss
.footer {
  // this theme has uses a less bright background color for its footer
  --color-background: var(--color-background-muted);
}
```

#### Text Colors

Colors for various types and states of text. Something to note here is that the definition of "text" is pretty flexible. Text color tokens can be used as defaults for any "text-like" things, such as inline icons, list bullets, etc. Basically anything that sits inline and needs to have contrast with the background.

### Link Colors

> A typical site will probably customize links in multiple places, but as defaults to use for building our components we provide two sets of link tokens. The defaults for links, `--color-link-*` and a second set to use in navigation elements, `--color-navigation-link-*`.

Reviewer: ```suggestion
A typical site will probably customize links in multiple places, but as defaults to use for building our components, we provide two sets of link tokens. The defaults for links, `--color-link-*` and a second set to use in navigation elements, `--color-navigation-link-*`.
```
